File: test/jsdom/consentblock.test.mjs
// P02.S03 — the signer sees WHERE their signature will land, on the screen they decide from.
//
// **The defect this drives.** The consent screen rendered every page of the received document and
// said nothing about the block. `handleSessionQuote`'s rect is `p2p.NominalBlockRect()`, whose own
// doc calls it *"a size template, not a placement — the caller wants a rect of the right shape and
// must not care where it says it is"*, and app.js consumed only its width and height. The real
// placement was computed server-side after the user had already consented.
//
// **What only this tier can see: the FLIP.** A PDF rect is in points with the origin at the bottom
// left; a canvas measures from the top left. Getting that wrong does not throw and does not fail a
// Go test — it draws a box, correctly sized, in the wrong half of the page, on the screen where a
// signer is being told where their signature goes. The Go side asserts the numbers are the ones
// that will be stamped; nothing there can see what they become on screen.
//
// One boot per file — see boot.mjs.
import test from 'node:test';
import assert from 'node:assert/strict';
import { boot } from './boot.mjs';
import { setNextDocument } from './stub-pdfjs.mjs';

const PEER = { fingerprint: 'a'.repeat(64), label: 'Ada' };
// A block low on page **2**, and both halves of that are deliberate.
//
// `lly: 40` in PDF points is near the BOTTOM of a 792pt page, which is what makes the flip
// observable: unflipped it renders near the top.
//
// **Page 2 rather than page 1 because a one-page fixture cannot see the page match at all** —
// measured: with `numPages: 1`, dropping `block.page === i` from the draw condition left this file
// green, because the condition it removed was always true. A two-page document makes the wrong
// answer land somewhere a test can point at.
const BLOCK = { page: 2, rect: [40, 40, 320, 124] };

let armed = false;
let withBlock = true;
const h = await boot({
  routes: {
    '/api/peers': () => ({ self: 'f'.repeat(64), peers: [PEER] }),
    '/api/session/arm': () => { armed = true; return { armed: true, address: '127.0.0.1:8443' }; },
    '/api/session/status': () => (armed
      ? {
        armed: true,
        address: '127.0.0.1:8443',
        pending: {
          signer: 'Ada Landlord',
          fingerprint: 'a'.repeat(64),
          reason: 'I agree to co-sign the lease',
          signers: [],
          // Absent, not empty, when the server could not compute one — the field is `omitempty`.
          ...(withBlock ? { block: BLOCK } : {}),
        },
      }
      : { armed: false }),
    '/api/session/disarm': () => { armed = false; return {}; },
  },
});
const { document: doc, settle } = h;

async function reachConsent() {
  // `renders: true` is required here and nowhere else: the consent preview's failure path replaces
  // the whole column with "could not render the document", so with the default stub the geometry
  // under test is wiped before it can be read. See stub-pdfjs.mjs.
  setNextDocument({ numPages: 2, renders: true });
  doc.getElementById('sessionRecvBtn').click();
  await settle();
  doc.getElementById('srvPeer').selectedIndex = 0;
  doc.getElementById('srvArmGo').click();
  await settle();
  await new Promise((r) => setTimeout(r, 1800));
  await settle();
}

async function cleanUp() {
  doc.getElementById('srvCancel').click();
  await settle();
}

test('the block is drawn on its own page, measured from the bottom of the page', async () => {
  withBlock = true;
  await reachConsent();
  const onConsent = !doc.getElementById('srvConsent').hidden;
  const pages = doc.querySelectorAll('#srvPreview .srvpage');
  const box = doc.querySelector('#srvPreview .srvblock');
  const geom = box ? { ...box.style } && {
    left: box.style.left, top: box.style.top, width: box.style.width, height: box.style.height,
  } : null;
  const onItsOwnPage = box ? box.parentElement === pages[1] : false;
  await cleanUp();

  // SETUP: without these the assertions below are satisfied by a screen that never opened or a
  // preview that rendered nothing.
  assert.ok(onConsent, 'the consent screen was never reached, so nothing below proves anything');
  assert.equal(pages.length, 2, 'the preview did not render both pages, so "the block is on '
    + 'page 2 and not page 1" is not a distinction this run could have drawn');
  assert.ok(box, 'the consent screen shows the document and says nothing about where this '
    + 'signature will land — the signer decides blind');
  assert.ok(onItsOwnPage, 'the block was drawn on the wrong page, or outside a page wrapper '
    + 'altogether — on the first it tells the signer their signature lands somewhere it does not, '
    + 'and on the second it is positioned against the scrolling column and slides off on the '
    + 'first scroll');

  // The scale: the preview fits pages to 380px, so 380/612 for a US-Letter page.
  const scale = 380 / 612;
  const pageHeight = 792 * scale;
  const top = parseFloat(geom.top);
  const left = parseFloat(geom.left);

  // THE FLIP. lly=40 sits near the bottom of the page, so the box's CSS top must be in the LOWER
  // half. Unflipped (`top = lly * scale`) it lands at ~25px, near the top — correctly sized, and
  // in the wrong place.
  assert.ok(top > pageHeight / 2,
    `the block is drawn ${top}px from the top of a ${pageHeight}px page, which is the upper half `
    + '— a PDF rect measures from the BOTTOM left and a canvas from the top left, so this is the '
    + 'flip being missed rather than the rect being wrong');
  assert.ok(Math.abs(top - (pageHeight - 124 * scale)) < 1,
    `the block's top is ${top}px, want about ${pageHeight - 124 * scale}px`);
  assert.ok(Math.abs(left - 40 * scale) < 1, `the block's left is ${left}px, want ${40 * scale}px`);
  assert.ok(Math.abs(parseFloat(geom.width) - 280 * scale) < 1,
    `the block is ${geom.width} wide, want ${280 * scale}px — the size is what the appearance `
    + 'image is rasterised to, so a box of the wrong size misreports what will be stamped');
  assert.ok(Math.abs(parseFloat(geom.height) - 84 * scale) < 1,
    `the block is ${geom.height} tall, want ${84 * scale}px`);
});

test('a document with no computable placement still shows its pages', async () => {
  withBlock = false;
  await reachConsent();
  const onConsent = !doc.getElementById('srvConsent').hidden;
  const pages = doc.querySelectorAll('#srvPreview .srvpage');
  const box = doc.querySelector('#srvPreview .srvblock');
  await cleanUp();

  assert.ok(onConsent, 'the consent screen was never reached, so nothing below proves anything');
  assert.equal(pages.length, 2, 'the preview stopped rendering pages when there was no block — '
    + 'the block is an annotation ON the review and must not be able to take the review with it');
  assert.equal(box, null, 'a box was drawn for a document whose placement the server could not '
    + 'compute, so the signer is being shown a location Nib does not have');
});

File: internal/server/consentblock_test.go
package server

import (
	"encoding/hex"
	"os"
	"strings"
	"testing"

	"nib/internal/p2p"
	"nib/internal/sign"
	"nib/internal/testpdf"
)

// P02.S03 — the signer sees where their block will land, on the surface they decide from.
//
// # The defect these drive
//
// `handleSessionQuote` answers with `p2p.NominalBlockRect()`, and its own doc says that is *"a size
// template, not a placement — the caller wants a rect of the right shape and must not care where it
// says it is"*; `web/app.js` consumes only its width and height. The real placement was computed
// server-side AFTER consent, at `p2p/session.go:1097`. So a party decided whether to sign, and only
// then did anything work out where their signature would go.

// TestTheBlockShownIsTheBlockStamped is the slice's second acceptance clause, and it is asserted as
// an EQUALITY between two calls rather than as a property of one.
//
// **The equality is the whole point and it is why this is not a comparison of two implementations.**
// `blockFor` and the stamp both call `p2p.PlacementFor`, ADR-009's one door for this question, and
// the inputs are the same values rather than equal ones: `Confirm` is handed `doc []byte`, which is
// `coSignExchange`'s `inbound`; the roster is `cer.l3Roster()` on the same `cer` the confirmer is
// built with (`p2p.Receive(` has one production call site and `sessionConfirmer{` one construction,
// and they are the same line). This test pins that the door is what `blockFor` reaches, so a future
// hand-copied placement fails here rather than drifting silently.
func TestTheBlockShownIsTheBlockStamped(t *testing.T) {
	doc, err := testpdf.Text("the lease")
	if err != nil {
		t.Fatal(err)
	}
	cert, _, err := sign.GenerateIdentity("Signer")
	if err != nil {
		t.Fatal(err)
	}
	myFP, err := sign.Fingerprint(cert)
	if err != nil {
		t.Fatal(err)
	}

	// The stamp side, called exactly as `coSignExchange` calls it: the received bytes, the arm's
	// roster, this machine's fingerprint. Outside a ceremony the roster is the zero value, which
	// is the manual co-sign path and which `PlacementFor` answers with `NextPlacement`.
	want, err := p2p.PlacementFor(doc, p2p.Roster{}, hex.EncodeToString(myFP))
	if err != nil {
		t.Fatal(err)
	}

	got := sessionConfirmer{}.blockFor(doc, myFP)
	if got == nil {
		t.Fatal("the consent view carries no block for a document whose placement computes fine, " +
			"so the signer decides without being shown where their signature lands")
	}
	if got.Page != want.Page || got.Rect != want.Rect {
		t.Errorf("the consent surface shows page %d %v and the signature is stamped at page %d %v "+
			"— a signer shown one location and given another is worse off than one shown none",
			got.Page, got.Rect, want.Page, want.Rect)
	}
	// SETUP, and it is the assertion that keeps the one above honest: a placement that was
	// hardcoded to the nominal rect would satisfy an equality against a nominal `want` too. The
	// nominal rect is a size template at index 0; a real placement on this document must not be it.
	if got.Rect == p2p.NominalBlockRect() && got.Page == 0 {
		t.Error("the reported block is the nominal size template rather than a placement — that " +
			"is the value this slice exists to stop sending")
	}
}

// TestAnUncomputableBlockLeavesTheRestOfTheSurface — STANDARDS §9, the rule `loadPendingPreview`
// already cites for the preview: a box that cannot be worked out degrades that box and not the
// screen where somebody decides whether to sign.
func TestAnUncomputableBlockLeavesTheRestOfTheSurface(t *testing.T) {
	myFP := []byte(strings.Repeat("\x01", 32))
	got := sessionConfirmer{}.blockFor([]byte("this is not a PDF"), myFP)
	if got != nil {
		t.Errorf("a document whose placement cannot be computed produced a block %v — a rect of "+
			"zeroes on page zero is a LOCATION, and this is the absence of one, so the field is "+
			"omitted rather than sent empty", got)
	}
}

// TestTheConsentViewSendsTheBlock is the routing half: the rule exists and the view calls it.
//
// It is a source scan because the behavioural reach needs a live session with a peer mid-handshake,
// which no tier below 4 arranges — the same limit `recitalFor`'s call-site scan records one field
// over. A scan proves the line is present; `TestTheBlockShownIsTheBlockStamped` proves it is right.
func TestTheConsentViewSendsTheBlock(t *testing.T) {
	src, err := os.ReadFile("session.go")
	if err != nil {
		t.Fatal(err)
	}
	code := string(src)
	body := funcBodyFrom(code, strings.Index(code, "func (sc sessionConfirmer) Confirm("))
	if !strings.Contains(body, "view.Block = sc.blockFor(") {
		t.Error("the consent view no longer carries the signer's own block. blockFor can be " +
			"perfectly correct and reach nobody: the field is `omitempty`, so a view that stops " +
			"setting it is indistinguishable at the client from a document with no placement")
	}
}
